src/setup_payload/ManualSetupPayloadGenerator.cpp

```diff
@@ -87,56 +87,78 @@ static uint32_t chunk3PayloadRepresentation(const SetupPayload & payload)
-    snprintf(buf, sizeof(buf), "%0*" PRIu32, minLength, number);
-    return std::string(buf);
+    snprintf(buf, len + 1, "%0*" PRIu32, static_cast<int>(len), number);
 }
 
-#pragma GCC diagnostic pop
-
-CHIP_ERROR ManualSetupPayloadGenerator::payloadDecimalStringRepresentation(std::string & outDecimalString)
+CHIP_ERROR ManualSetupPayloadGenerator::payloadDecimalStringRepresentation(MutableCharSpan & outBuffer)
 {
-    if (!mSetupPayload.isValidManualCode())
-    {
-        ChipLogError(SetupPayload, "Failed encoding invalid payload");
-        return CHIP_ERROR_INVALID_ARGUMENT;
-    }
-
     static_assert(kManualSetupCodeChunk1CharLength + kManualSetupCodeChunk2CharLength + kManualSetupCodeChunk3CharLength ==
                       kManualSetupShortCodeCharLength,
-                  "Manual code length mismatch");
+                  "Manual code length mismatch (short)");
+    static_assert(kManualSetupShortCodeCharLength + kManualSetupVendorIdCharLength + kManualSetupProductIdCharLength ==
+                      kManualSetupLongCodeCharLength,
+                  "Manual code length mismatch (long)");
     static_assert(kManualSetupChunk1DiscriminatorMsbitsLength + kManualSetupChunk2DiscriminatorLsbitsLength ==
                       kManualSetupDiscriminatorFieldLengthInBits,
                   "Discriminator won't fit");
     static_assert(kManualSetupChunk2PINCodeLsbitsLength + kManualSetupChunk3PINCodeMsbitsLength == kSetupPINCodeFieldLengthInBits,
                   "PIN code won't fit");
 
-    uint32_t chunk1 = chunk1PayloadRepresentation(mSetupPayload);
-    uint32_t chunk2 = chunk2PayloadRepresentation(mSetupPayload);
-    uint32_t chunk3 = chunk3PayloadRepresentation(mSetupPayload);
+    if (!mPayloadContents.isValidManualCode())
+    {
+        ChipLogError(SetupPayload, "Failed encoding invalid payload");
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    bool useLongCode = (mPayloadContents.commissioningFlow == CommissioningFlow::kCustom);
 
-    std::string decimalString = decimalStringWithPadding(chunk1, kManualSetupCodeChunk1CharLength);
-    decimalString += decimalStringWithPadding(chunk2, kManualSetupCodeChunk2CharLength);
-    decimalString += decimalStringWithPadding(chunk3, kManualSetupCodeChunk3CharLength);
+    if ((useLongCode && outBuffer.size() < kManualSetupLongCodeCharLength + 2) ||
+        (!useLongCode && outBuffer.size() < kManualSetupShortCodeCharLength + 2))
+    {
+        ChipLogError(SetupPayload, "Failed encoding payload to buffer");
+        return CHIP_ERROR_BUFFER_TOO_SMALL;
+    }
+
+    uint32_t chunk1 = chunk1PayloadRepresentation(mPayloadContents);
+    uint32_t chunk2 = chunk2PayloadRepresentation(mPayloadContents);
+    uint32_t chunk3 = chunk3PayloadRepresentation(mPayloadContents);
+
+    char * stringPtr = outBuffer.data();
 
-    if (mSetupPayload.commissioningFlow == CommissioningFlow::kCustom)
+    decimalStringWithPadding(stringPtr, kManualSetupCodeChunk1CharLength, chunk1);
```

Comment:
you should check outBuffer.size() to validate how much space can be written in these places.

can decimalStringWithPadding return a status (like return error if insufficient space)?

Reply:
Added space checks.